sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -61,6 +61,9 @@ case class InMemoryTableScanExec(
     }) && !WholeStageCodegenExec.isTooManyFields(conf, relation.schema)
   }
 
+  // TODO: revisit this. Shall we always turn off whole stage codegen if the output data are rows?
+  override def supportCodegen: Boolean = supportsBatch
```

Comment:
In 2.4 we should look into this. My gut feeling is we don't need to enable whole stage codegen for scan nodes that output data as rows.

Reply:
Yeah, we can do more perf measurement after 2.3 release

Reply:
I think this is safe to keep the same behavior of 2.2.

I'm not sure if enabling whole stage codegen can hurt performance for scan nodes, btw. We can revisit this, of course.

Reply:
We confirmed that we got performance improvements in several cases.
I will revisit this with more cases after 2.3 release.